core/src/qrcode/MQRDecoder.cpp

```diff
@@ -0,0 +1,392 @@
+	return bits.peakBits(bitsAvailable) == 0;
+}
+
+/**
+ * <p>QR Codes can encode text as bits in one of several modes, and can use multiple modes
+ * in one QR Code. This method decodes the bits back into text.</p>
+ *
+ * <p>See ISO 18004:2006, 6.4.3 - 6.4.7</p>
+ */
+ZXING_EXPORT_TEST_ONLY DecoderResult DecodeBitStream(ByteArray&& bytes, const Version& version,
+													 QRCode::ErrorCorrectionLevel ecLevel,
+													 const std::string& hintedCharset)
+{
+	BitSource bits(bytes);
+	std::wstring result;
+	static const int GB2312_SUBSET = 1;
+	const int modeBitLength = CodecModeBitsLength(version);
+	const int minimumBitsRequired = modeBitLength + CharacterCountBits(CodecMode::NUMERIC, version);
+	try {
+		CharacterSet currentCharset = CharacterSet::Unknown;
+		CodecMode mode;
+		do {
+			// While still another segment to read...
+			if (bits.available() < minimumBitsRequired || IsTerminator(bits, version)) {
+				// OK, assume we're done. Really, a TERMINATOR mode should have been recorded here
+				mode = CodecMode::TERMINATOR;
+			} else if (modeBitLength == 0 && version.versionNumber() == 1) {
+				mode = CodecMode::NUMERIC;
+			} else {
+				mode = CodecModeForBits(bits.readBits(modeBitLength), version); // mode is encoded by 4 bits
+			}
+			switch (mode) {
+			case CodecMode::TERMINATOR: break;
+			case CodecMode::HANZI: {
+				// First handle Hanzi mode which does not start with character count
+				// chinese mode contains a sub set indicator right after mode indicator
+				int subset = bits.readBits(4);
+				int countHanzi = bits.readBits(CharacterCountBits(mode, version));
+				if (subset == GB2312_SUBSET) {
+					auto status = DecodeHanziSegment(bits, countHanzi, result);
+					if (StatusIsError(status)) {
+						return status;
+					}
+				}
+				break;
+			}
+			default: {
+				// "Normal" QR code modes:
+				// How many characters will follow, encoded in this mode?
+				int count = bits.readBits(CharacterCountBits(mode, version));
+				DecodeStatus status;
+				switch (mode) {
+				case CodecMode::NUMERIC: status = DecodeNumericSegment(bits, count, result); break;
+				case CodecMode::ALPHANUMERIC: status = DecodeAlphanumericSegment(bits, count, false, result); break;
+				case CodecMode::BYTE:
+					status = DecodeByteSegment(bits, count, currentCharset, hintedCharset, result);
+					break;
+				case CodecMode::KANJI: status = DecodeKanjiSegment(bits, count, result); break;
+				default: status = DecodeStatus::FormatError;
+				}
+				if (StatusIsError(status)) {
+					return status;
+				}
+				break;
+			}
+			}
+		} while (mode != CodecMode::TERMINATOR);
+	} catch (const std::exception&) {
+		// from readBits() calls
+		return DecodeStatus::FormatError;
+	}
+
+	return DecoderResult(std::move(bytes), std::move(result)).setEcLevel(ToString(ecLevel));
+}
+
+static DecoderResult DoDecode(const BitMatrix& bits, const Version& version, const std::string& hintedCharset,
```

Comment:
Also the same code except for the different implementation of the `ReadFormatInformation` and other helper functions. I think a bool flag either passed to those or used to descriminate between calling them from either of the two namespaces should be the simplest way to avoid the code duplication?